DigitalLearningSolutions.Web.Tests/Controllers/Logout/LogoutControllerTests.cs

```diff
@@ -0,0 +1,60 @@
+﻿namespace DigitalLearningSolutions.Web.Tests.Controllers.Logout
+{
+    using System;
+    using System.Security.Claims;
+    using System.Threading.Tasks;
+    using DigitalLearningSolutions.Data.Tests.Helpers;
+    using DigitalLearningSolutions.Web.Controllers;
+    using FakeItEasy;
+    using FluentAssertions.AspNetCore.Mvc;
+    using Microsoft.AspNetCore.Authentication;
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.AspNetCore.Mvc.Routing;
+    using NUnit.Framework;
+
+    public class LogoutControllerTests
+    {
+        private LogoutController controller;
+
+        [SetUp]
+        public void SetUp()
+        {
+            // Set up authenticated user
+            var user = new ClaimsPrincipal(new ClaimsIdentity("mock"));
+            var session = new MockHttpContextSession();
+
+            var authService = A.Fake<IAuthenticationService>();
+            A.CallTo(() => authService.SignInAsync(A<HttpContext>._, A<string>._, A<ClaimsPrincipal>._,
+                A<AuthenticationProperties>._)).Returns(Task.CompletedTask);
+
+            var urlHelperFactory = A.Fake<IUrlHelperFactory>();
+            var services = A.Fake<IServiceProvider>();
+            A.CallTo(() => services.GetService(typeof(IAuthenticationService))).Returns(authService);
+            A.CallTo(() => services.GetService(typeof(IUrlHelperFactory))).Returns(urlHelperFactory);
+
+            controller = new LogoutController
+            {
+                ControllerContext = new ControllerContext
+                {
+                    HttpContext = new DefaultHttpContext
+                    {
+                        User = user,
+                        Session = session,
+                        RequestServices = services
+                    }
+                }
+            };
+        }
+
+        [Test]
+        public void Logout_should_redirect_user_to_home_page()
```

Comment:
Would it be feasible to also have a test that the intended logout has actually occurred? Right now it'd be possible to pass the tests without the logout controller method changing anything.

Reply:
It doesn't look like there is an easy way to do this unfortunately. The SignOutAsync method is a static extension method that can't be intercepted, and due to the way that the authentication service is mocked, the User Identity details aren't accurately represented in tests. I tried having a look into incorporating something similar into the integration tests, but there seems to be similar issues with the authentication service.

Reply:
Oh, I meant more if the authService that we mock when creating a controller gets called as a result of this? (I think it has SignOutAsync method on it that we mock explicitly)

Reply:
Oh sorry, it's in ControllerContextHelper currently

Reply:
Yeah the set up is all a bit confusing, but the mocked authService all gets bundled up together and put in to the HttpContext, which is where it is accessed from. 

Might be able to pull that out into the ControllerTests, I'll take a look.

Reply:
You were right, I didn't even think of this approach but it works perfectly. I implemented a test that ensures that the SignOutAsync method is called when logging out, and did a similar test in the LoginControllerTests for SignInAsync.